src/main/kotlin/me/urielsalis/mccrashlib/deobfuscator/Deobfuscator.kt

```diff
@@ -3,54 +3,72 @@ package me.urielsalis.mccrashlib.deobfuscator
 import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
 import proguard.retrace.ReTrace
 import java.io.File
-import java.io.FileOutputStream
+import java.io.IOException
 import java.io.LineNumberReader
 import java.io.PrintWriter
 import java.io.StringReader
 import java.io.StringWriter
-import java.net.HttpURLConnection
+import java.net.URI
 import java.net.URL
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.time.Duration
 
-fun getDeobfuscation(modded: Boolean, version: String, content: String, isClient: Boolean, mappingsDirectory: File): String? {
-    if (modded) {
+private fun getMappingFile(mappingsDirectory: File, version: String, isClient: Boolean): File? {
+    // Reject malformed or malicious names
+    if (version.isBlank() || version.contains("\\") || version.contains("/")) {
         return null
     }
 
-    if (version.isBlank() || version.contains("\\") || version.contains("/") || isZipSlip(version, mappingsDirectory)) {
+    val name = if (isClient) {
+        "$version-client.txt"
+    } else {
+        "$version-server.txt"
+    }
+
+    val canonicalDestinationDir = mappingsDirectory.canonicalPath
+    val destinationFile = File(mappingsDirectory, name)
+    val canonicalDestinationFile = destinationFile.canonicalPath
+    // Reject if file would be outside intended directory
+    // Based on Zip Slip mitigation, see https://snyk.io/research/zip-slip-vulnerability#java
+    if (!canonicalDestinationFile.startsWith(canonicalDestinationDir + File.separator)) {
         return null
     }
+    return destinationFile
+}
 
-    val name = if (isClient) {
-        "$version-client"
-    } else {
-        "$version-server"
+fun getDeobfuscation(
+    modded: Boolean,
+    version: String,
+    content: String,
+    isClient: Boolean,
+    mappingsDirectory: File
+): String? {
+    if (modded) {
+        return null
     }
 
-    val mappingFile = synchronized(mappingsDirectory) {
-        val mappingFile = File(mappingsDirectory, name)
+    val mappingFile = getMappingFile(mappingsDirectory, version, isClient) ?: return null
+
+    // Synchronize to prevent race condition while downloading mapping file
+    synchronized(mappingsDirectory) {
         if (!mappingFile.exists()) {
-            downloadMapping(version, name, isClient, mappingsDirectory)
+            downloadMapping(version, mappingFile, isClient)
         }
         if (!mappingFile.exists()) {
             return null
         }
-        mappingFile
     }
     val retrace = ReTrace(ReTrace.REGULAR_EXPRESSION, false, true, mappingFile)
     val stringWriter = StringWriter()
     val printWriter = PrintWriter(stringWriter)
     retrace.retrace(LineNumberReader(StringReader(content)), printWriter)
+    printWriter.flush()
```

Comment:
This is not strictly required because `ReTrace` already does it, but that is an implementation detail and it probably does not hurt to have it here as well.